src/WebJobs.Script/Host/HostFunctionMetadataProvider.cs

```diff
@@ -179,10 +183,21 @@ internal static string ParseLanguage(string scriptFilePath, IEnumerable<RpcWorke
 
             // determine the script type based on the primary script file extension
             string extension = Path.GetExtension(scriptFilePath).ToLowerInvariant().TrimStart('.');
-            var workerConfig = workerConfigs.FirstOrDefault(config => config.Description.Extensions.Contains("." + extension));
-            if (workerConfig != null)
+
+            // In the special case of a null FUNCTIONS_WORKER_RUNTIME and an extension of "dll",
+            // we want to skip the worker config check (to prevent the dotnet-isolated worker from
+            // being chosen) and allow in-proc handling to be used.
+            // Other notes:
+            // - when dotnet-isolated, the FUNCTIONS_WORKER_RUNTIME is required to be set to "dotnet-isolated"
+            // - when FUNCTIONS_WORKER_RUNTIME is set to "dotnet", workerConfigs is empty so this correctly falls
+            //   through to choose DotNetAssembly with a "dll" extension
+            if (!(functionsWorkerRuntime is null && extension == "dll"))
             {
-                return workerConfig.Description.Language;
+                var workerConfig = workerConfigs.FirstOrDefault(config => config.Description.Extensions.Contains("." + extension));
+                if (workerConfig != null)
+                {
+                    return workerConfig.Description.Language;
+                }
```

Comment:
The fix is here -- hopefully this description is good enough to explain to us in the future why it works this way. Eventually we'll have a no-in-proc host so this flow won't matter as much.